validator/client/validator_propose.go

```diff
@@ -0,0 +1,77 @@
+package client
+
+// Validator client proposer functions.
+
+import (
+	"context"
+
+	ptypes "github.com/gogo/protobuf/types"
+	"github.com/opentracing/opentracing-go"
+	pbp2p "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/ssz"
+)
+
+// ProposeBlock
+//
+// WIP - not done.
+func (v *validator) ProposeBlock(ctx context.Context, slot uint64) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "validator.ProposeBlock")
+	defer span.Finish()
+
+	// 1. Fetch data from Beacon Chain node.
+	// Get current head beacon block.
+	headBlock, err := v.beaconClient.CanonicalHead(ctx, &ptypes.Empty{})
+	if err != nil {
+		log.Errorf("Failed to fetch CanonicalHead: %v", err)
+		return
+	}
+	_ = headBlock // TODO(1461): Actually tree hash the block.
+	parentTreeHash, err := ssz.TreeHash(false)
+	if err != nil {
+		log.Errorf("Failed to hash parent block: %v", err)
+		return
+	}
+
+	// Get pending deposit data.
+	pDepResp, err := v.beaconClient.PendingDeposits(ctx, &ptypes.Empty{})
+	if err != nil {
+		log.Errorf("Failed to get pending pendings: %v", err)
+		return
+	}
+
+	// Get ETH1 data.
+	eth1DataResp, err := v.beaconClient.Eth1Data(ctx, &ptypes.Empty{})
+	if err != nil {
+		log.Errorf("Failed to get ETH1 data: %v", err)
+		return
+	}
+
+	// 2. Construct block
+	block := &pbp2p.BeaconBlock{
+		Slot:               slot,
+		ParentRootHash32:   parentTreeHash[:],
+		RandaoRevealHash32: nil, // TODO(1366): generate randao reveal from BLS
+		Eth1Data:           eth1DataResp.Eth1Data,
+		Body: &pbp2p.BeaconBlockBody{
+			Attestations:      v.attestationPool.PendingAttestations(),
+			ProposerSlashings: nil, // TODO(1438): Add after operations pool
+			AttesterSlashings: nil, // TODO(1438): Add after operations pool
+			Deposits:          pDepResp.PendingDeposits,
+			Exits:             nil, // TODO(1323): Add validator exits
+		},
+	}
+
+	// 3. Compute state root transition from parent block to the new block
+	resp, err := v.proposerClient.ComputeStateRoot(ctx, block)
+	if err != nil {
+		log.Errorf("Unable to compute state root: %v", err)
+	}
+	block.StateRootHash32 = resp.GetStateRoot()
+
+	// 4. Sign the complete block
+	// TODO(1366): BLS sign block
+	block.Signature = nil
+
+	// 5. Broadcast to the network
+	v.p2p.Broadcast(block)
```

Comment:
As discussed in discord, do this in the chain service after block being received